admin/provisioner/clickhousestatic/provisioner.go

```diff
@@ -102,12 +102,20 @@ func (p *Provisioner) Provision(ctx context.Context, r *provisioner.Resource, op
 		return nil, fmt.Errorf("failed to create clickhouse database: %w", err)
 	}
 
-	// Idempotently create the user
+	// Idempotently create the user.
 	_, err = p.ch.ExecContext(ctx, fmt.Sprintf("CREATE USER IF NOT EXISTS %s IDENTIFIED WITH sha256_password BY ? DEFAULT DATABASE %s GRANTEES NONE", user, dbName), password)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create clickhouse user: %w", err)
 	}
 
+	// When creating the user, the password assignment is not idempotent (if there are two concurrent invocations, we don't know which password was used).
+	// By adding the password separately, we ensure all passwords will work.
+	// NOTE: Requires ClickHouse 24.9 or later.
+	_, err = p.ch.ExecContext(ctx, fmt.Sprintf("ALTER USER %s ADD IDENTIFIED WITH sha256_password BY ?", user), password)
```

Comment:
Seems risky, in a rare case of id collision, would it end up giving an existing project access to previously created DB ? However, in case of id collision we would have even bigger troubles..
Usually, locking on table name would be less prone to errors...

Reply:
Thanks for taking a look – this should not be a problem because:
1. The risk of an ID collision is the same as the risk of guessing the password (both have 16 bytes of randomness), so extremely unlikely.
2. Even if there's an ID collision, the resource is inserted in our Postgres before we provision the ClickHouse account. This will cause a unique constraint error on the ID in Postgres and ClickHouse will not be provisioned until the next retry (which will generate a new ID).